src/VisualStudio/Core/Next/FindReferences/StreamingFindUsagesPresenter.cs

```diff
@@ -47,34 +51,104 @@ internal partial class StreamingFindUsagesPresenter :
-            _typeMap = typeMap;
-            _formatMapService = formatMapService;
+            TextEditorFactoryService = textEditorFactoryService;
+            TypeMap = typeMap;
+            FormatMapService = formatMapService;
 
             _vsFindAllReferencesService = (IFindAllReferencesService)_serviceProvider.GetService(typeof(SVsFindAllReferences));
         }
 
-        public FindUsagesContext StartSearch(string title, bool alwaysShowDeclarations)
+        public FindUsagesContext StartSearch(string title, bool supportsReferences)
         {
             this.AssertIsForeground();
 
             // Get the appropriate window for FAR results to go into.
             var window = _vsFindAllReferencesService.StartSearch(title);
 
-            // Make the data source that will feed data into this window.
-            var dataSource = new TableDataSourceFindUsagesContext(
-                this, window, alwaysShowDeclarations);
+            // Keep track of the users preference for grouping by definition if we don't already know it.
+            // We need this because we disable the Definition column when wer're not showing references
+            // (i.e. GoToImplementation/GoToDef).  However, we want to restore the user's choice if they
+            // then do another FindAllReferences.
+            var desiredGroupingPriority = _workspace.Options.GetOption(FindUsagesOptions.DefinitionGroupingPriority);
+            if (desiredGroupingPriority < 0)
+            {
+                StoreCurrentGroupingPriority(window);
+            }
 
-            // And return the data source so that the FindRefs engine can report results
-            // which the data source can then create the appropriate presentation items for
-            // for the window.
-            return dataSource;
+            return supportsReferences
+                ? StartSearchWithReferences(window, desiredGroupingPriority)
+                : StartSearchWithoutReferences(window);
+        }
+
+        private FindUsagesContext StartSearchWithReferences(IFindAllReferencesWindow window, int desiredGroupingPriority)
+        {
+            // Ensure that the window's definition-grouping reflects what the user wants.
+            // i.e. we may have disabled this column for a previous GoToImplementation call. 
+            // We want to still show the column as long as the user has not disabled it themselves.
+            var definitionColumn = window.GetDefinitionColumn();
+            if (definitionColumn.GroupingPriority != desiredGroupingPriority)
+            {
+                SetDefinitionGroupingPriority(window, desiredGroupingPriority);
+            }
+
+            // If the user changes the grouping, then store their current preference.
+            var tableControl = (IWpfTableControl2)window.TableControl;
+            tableControl.GroupingsChanged += (s, e) => StoreCurrentGroupingPriority(window);
```

Comment:
Isn't this going to create and hook up a new delegate every time the method is called?